### PR TITLE
ledger state machine enhancements

### LINKED ISSUE #2219: ledger panel not resetting when switching chain
**Wallet version or commit SHA:**
7f262913c98eee5c30a55957ad2c565d733c5845

**Environment:**
linux/osx/windows

**Prerequisites:**
have a ledger

**Reproduction steps:**
e.g.:
1. open Ledger tab in N2 mode
2. open the ledger app such that we get an address
3. now switch to N3 mode. The data in the panel will not be refreshed and shows addresses for the wrong chain.

Same happens if you start with N3 and go back to N2.

### PR DESCRIPTION
**Pull request by lllwvlvwlll**
This PR resolves #2219 by introducing a number of enhancements:
1. Continues to monitor and update the ledger component once a ledger app has initially opened
2. Removes caching of the address list after the ledger app has closed
3. Prevents use of the Neo Legacy Ledger app on Neo N3 and vice-versa

### CODE REVIEW

**Comment by ixje** [comment]
The build was failing because the `node-hid` packaged change their release tool to `n-api` which creates different release names then what `electron-builder` was searching for. The fix is to upgrade `electron-builder` to at least `22.11.4`, we updated to `22.11.11` (latest stable). This in turn creates a new problem where it fails to import `fs/promises`. To fix that we need to upgrade `node` to `>= 14.17.0` (According to this commit
 https://github.com/electron-userland/electron-builder/commit/96d7813362d319c3141d78e179c3f6d71c798fab). The new `wine-chrome` image uses node `14.17.0`

Note there was no lower `electron-builder` version that fixes the `node-hid` issue without requiring a newer `node` version.

File: package.json [hunk 1]
```diff
@@ -177,7 +177,7 @@
     "cross-env": "5.1.1",
     "css-loader": "0.28.7",
     "electron": "8.5.5",
-    "electron-builder": "22.9.1",
+    "electron-builder": "22.11.11",
```
Comment (comountainclimber) [inline package.json:180]: Whats up with this change? Have we verified this doesnt break anything on our builds/OTA updates

File: .circleci/config.yml [hunk 2]
```diff
@@ -3,7 +3,7 @@ version: 2
 references:
   container_config: &container_config
     docker:
-      - image: electronuserland/builder:wine-chrome-05.18
+      - image: electronuserland/builder:wine-chrome-05.21
```
Comment (comountainclimber) [inline .circleci/config.yml:6]: Have you verified that this works?

File: app/ledger/n3NeonLedger.js [hunk 3]
```diff
@@ -69,8 +69,11 @@ export default class NeonLedger3 {
     const paths = await NeonLedger3.list()
     if (paths.length === 0) throw new Error(MESSAGES.NOT_CONNECTED)
     if (paths[0]) {
-      const ledger = new NeonLedger3(paths[0])
-      return ledger.open()
+      let ledger = new NeonLedger3(paths[0])
+      ledger = await ledger.open()
+      const appName = await ledger.getAppName()
+      if (appName !== 'NEO3') throw new Error(MESSAGES.APP_CLOSED)
```
Comment (comountainclimber) [inline app/ledger/n3NeonLedger.js:75]: 👍

File: app/ledger/n3NeonLedger.js [hunk 4]
```diff
@@ -198,9 +209,10 @@ export const getStartInfo = async () => {
   const ledger = await NeonLedger3.init()
 
   try {
-    const deviceInfo = await ledger.getDeviceInfo()
-    const publicKey = await ledger.getPublicKey()
-    return { deviceInfo, publicKey }
+    return await Promise.all([
+      ledger.getDeviceInfo(),
+      ledger.getPublicKey(),
+    ]).then(([deviceInfo, publicKey]) => ({ deviceInfo, publicKey }))
```
Comment (comountainclimber) [inline app/ledger/n3NeonLedger.js]: `const [deviceInfo, publicKey] = await Promise.all(...)`

File: app/ledger/neonLedger.js [hunk 5]
```diff
@@ -76,8 +77,11 @@ export default class NeonLedger {
     const paths = await NeonLedger.list()
     if (paths.length === 0) throw new Error(MESSAGES.NOT_CONNECTED)
     if (paths[0]) {
-      const ledger = new NeonLedger(paths[0])
-      return ledger.open()
+      let ledger = new NeonLedger(paths[0])
```
Comment (comountainclimber) [inline app/ledger/neonLedger.js]: If we are not reassigning `ledger` why is it being instantiated with `let`?

File: app/ledger/neonLedger.js [hunk 6]
```diff
@@ -252,9 +265,10 @@ export const getPublicKeys = async (
 export const getDeviceInfo = async () => {
   const ledger = await NeonLedger.init()
   try {
-    const deviceInfo = await ledger.getDeviceInfo()
-    const publicKey = await ledger.getPublicKey()
-    return { deviceInfo, publicKey }
+    return await Promise.all([
+      ledger.getDeviceInfo(),
+      ledger.getPublicKey(),
+    ]).then(([deviceInfo, publicKey]) => ({ deviceInfo, publicKey }))
```
Comment (comountainclimber) [inline app/ledger/neonLedger.js]: See comment above about cleaning this up

File: package.json (hunk 1 above)
Comment (ixje) [reply package.json:180]: See this https://github.com/CityOfZion/neon-wallet/pull/2227#issuecomment-913486861 for the reason. I don't know how to verify builds/OTA. I just saw a failing Windows build on the PR and tried to help by resolving it.

File: .circleci/config.yml (hunk 2 above)
Comment (ixje) [reply .circleci/config.yml:6]: See this https://github.com/CityOfZion/neon-wallet/pull/2227#issuecomment-913486861
without this change the windows build/deploy does not work.

File: app/ledger/n3NeonLedger.js (hunk 4 above)
Comment (lllwvlvwlll) [reply app/ledger/n3NeonLedger.js]: resolved in b49c341178a97650cf73ab078f3d2b95fb16e521

File: app/ledger/neonLedger.js (hunk 6 above)
Comment (lllwvlvwlll) [reply app/ledger/neonLedger.js]: resolved in b49c341178a97650cf73ab078f3d2b95fb16e521

File: app/ledger/neonLedger.js (hunk 5 above)
Comment (lllwvlvwlll) [reply app/ledger/neonLedger.js]: resolved in d62e32b50d6708e2f72e8b6000ea81f659058a15

**Review by comountainclimber — APPROVED** [review]

**Pull request merged**